debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotServerReader.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.connector.mysql;
+
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.apache.kafka.connect.source.SourceRecord;
+
+import io.debezium.function.BufferedBlockingConsumer;
+import io.debezium.function.Predicates;
+import io.debezium.jdbc.JdbcConnection;
+import io.debezium.relational.TableId;
+import io.debezium.util.Clock;
+import io.debezium.util.Strings;
+
+/**
+ * A component that performs a snapshot of a MySQL server, and records the schema changes in {@link MySqlSchema}.
+ * 
+ * @author Randall Hauch
+ */
+public class SnapshotServerReader extends AbstractSnapshotReader {
+
+    private final boolean includeData;
+    private final SnapshotReaderMetrics metrics;
+    private final SnapshotMySQLUtility snapshotMySQLUtility;
+
+    /**
+     * Create a snapshot reader.
+     * 
+     * @param name the name of this reader; may not be null
+     * @param context the task context in which this reader is running; may not be null
+     */
+    public SnapshotServerReader(String name, MySqlTaskContext context) {
+        super(name, context);
+        this.includeData = !context.isSchemaOnlySnapshot();
+        metrics = new SnapshotReaderMetrics(context.clock());
+        metrics.register(context, logger);
+        snapshotMySQLUtility = new SnapshotMySQLUtility(context);
+    }
+
+    /**
+     * Perform the snapshot using the same logic as the "mysqldump" utility.
+     */
+    protected void execute() {
+        context.configureLoggingContext("snapshot");
+        final AtomicReference<String> sql = new AtomicReference<>();
+        final JdbcConnection mysql = context.jdbc();
+        final MySqlSchema schema = context.dbSchema();
+        final SourceInfo source = context.source();
+        final Clock clock = context.clock();
+        final long ts = clock.currentTimeInMillis();
+        logger.info("Starting snapshot for {} with user '{}'", context.connectionString(), mysql.username());
+        logRolesForCurrentUser(mysql);
+        logServerInformation(mysql);
+        try {
+            metrics.startSnapshot();
+
+            snapshotMySQLUtility.beginTransaction();
+
+            try {
+                // I can't help but feel like there's got to be a better way to do this.
+                if (!isRunning()) return;
+
+                snapshotMySQLUtility.lockAllTables();
+
+                if (!isRunning()) return;
+                //logger.info("\tsnapshot continuing with database(s): {}", includedDatabaseNames);
+
+                snapshotMySQLUtility.maybeReadTableList();
```

Comment:
This is slightly confusing - it should be clear that readTableList was executed as a part of lockAllTables whel global locak failed. How about passing two lambdas - one executed when global lock succeeds and the other one when it fails?

Reply:
Changed to explicitly read the table list iff the global lock succeed, and some comments about why.